### PR TITLE
feat(client): ability to ignore collection order during comparison

### LINKED ISSUE #32: Add ability to ignore list order when comparing objects
### Description

Currently, Lists with the same elements in a different order are considered changed and therefore audited. For my team's use case, this change is not meaningful and adds clutter to our audits. We would like to be able to ignore order when comparing lists.

### Proposed solution

add an ignoreListOrder boolean to AuditorEventConfig. add a method to ObjectDiffCheckerService to use instead of the simple previousValue != updatedValue check, something like this:

image

using hashCodes to sort and compare means it doesn't matter what type of element is in the list

### Alternatives considered

this could be the default, but some users might care about list order so i think it's better to have it be configurable. i also thought about trying to let users pass in a comparison method through the configuration, but that would leave more room for error and may not be needed.

### PR DESCRIPTION
**Pull request by bettyand**
# Description

Some internal users of this app want to ignore the order of lists and sets which convert to arrays during the comparison. Order-only changes are not considered meaningful for their use cases and add clutter to the logs.

I have added configuration to achieve this by using a value other than index when building paths to leaf nodes. For objects, the default is to use the "id" field since that is commonly used, but users can also pass in field name(s) to use instead if their  objects use a different unique field or combination of fields. For primitives, the value itself is used to build the path and in case of duplication in a collection, the app checks the number of instances to determine if change has occurred.

Misc: Explanatory comments, tests to cover new feature, confirmed existing tests all still pass. A little refactoring for readability while I worked, plus some typo/formatting/linter fixes.

Fixes #32 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have added examples for the new scenarios in the relevant modules < **_unsure if this is necessary, requires configuration level changes_**

### CODE REVIEW

File: client/src/main/kotlin/com/lowes/auditor/client/entities/domain/AuditorEventConfig.kt [hunk 1]
```diff
@@ -154,3 +156,16 @@ data class RetryPublisherConfig(
     var count: Long? = null,
     var delay: Duration? = null
 )
+
+/**
+ * Data class configuration for ignoring order when comparing collection elements.
+ * This is done by using an identifier other than numerical index to build paths to values.
```
Comment (ankur4u007) [inline client/src/main/kotlin/com/lowes/auditor/client/entities/domain/AuditorEventConfig.kt]: Can we change it to `unique identifier`

File: client/src/test/kotlin/com/lowes/auditor/client/infrastructure/frameworks/service/ObjectDiffCheckerServiceTest.kt [hunk 2]
```diff
@@ -2,6 +2,7 @@ package com.lowes.auditor.client.infrastructure.frameworks.service
 
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.lowes.auditor.client.entities.domain.AuditorEventConfig
+import com.lowes.auditor.client.entities.domain.IgnoreCollectionOrderConfig
```
Comment (ankur4u007) [inline client/src/test/kotlin/com/lowes/auditor/client/infrastructure/frameworks/service/ObjectDiffCheckerServiceTest.kt:5]: Add a test for conditions when there is a duplicate entry in a list object.

File: client/src/test/kotlin/com/lowes/auditor/client/infrastructure/frameworks/service/ObjectDiffCheckerServiceTest.kt [hunk 3]
```diff
@@ -265,4 +311,111 @@ class ObjectDiffCheckerServiceTest : BehaviorSpec({
             }
         }
     }
+
+    Given("Diff Checker is configured to use alternate array identifier") {
```
Comment (ankur4u007) [inline client/src/test/kotlin/com/lowes/auditor/client/infrastructure/frameworks/service/ObjectDiffCheckerServiceTest.kt:270]: Add one additional test for nested list. Try with atleast 3 nested listed.
